features/FEATURE_LWIP/lwip-interface/ppp_lwip.cpp

```diff
@@ -0,0 +1,322 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include "platform/FileHandle.h"
+#include "platform/mbed_poll.h"
+#include "events/EventQueue.h"
+#include "rtos/Thread.h"
+#include "lwip/tcpip.h"
+#include "lwip/tcp.h"
+#include "lwip/ip.h"
+#include "lwip/dns.h"
+#include "lwip/pbuf.h"
+extern "C" { // "pppos.h" is missing extern C
+#include "netif/ppp/pppapi.h"
+}
+
+#include "lwip_stack.h"
+
+namespace mbed {
+
+using rtos::Thread;
+using events::EventQueue;
+
+#if LWIP_PPP_API
+
+static EventQueue *event_queue;
+static Thread *event_thread;
+static volatile bool event_queued;
+
+// Just one interface for now
+static ppp_pcb *my_ppp_pcb;
+
+static EventQueue *prepare_event_queue()
+{
+    if (event_queue) {
+        return event_queue;
+    }
+
+    // Should be trying to get a global shared event queue here!
+    // Shouldn't have to be making a private thread!
+
+    // Only need to queue 1 event. new blows on failure.
+    event_queue = new EventQueue(2 * EVENTS_EVENT_SIZE, NULL);
+    event_thread = new Thread(osPriorityNormal, 700);
+
+    if (event_thread->start(callback(event_queue, &EventQueue::dispatch_forever)) != osOK) {
+        delete event_thread;
+        delete event_queue;
+        return NULL;
+    }
+
+    return event_queue;
+}
+
+static u32_t ppp_output(ppp_pcb *pcb, u8_t *data, u32_t len, void *ctx)
+{
+    FileHandle *stream = static_cast<FileHandle *>(ctx);
+    pollfh fhs;
+    fhs.fh = stream;
+    fhs.events = POLLOUT;
+
+    // LWIP expects us to block on write
+    // File handle will be in non-blocking mode, because of read events.
+    // Therefore must use poll to achieve the necessary block for writing.
+
+    uint32_t written = 0;
+    while (len != 0) {
+        // Block forever until we're selected - don't care about reason we wake;
+        // return from write should tell us what's up.
+        poll(&fhs, 1, -1);
+        // This write will be non-blocking, but blocking would be fine.
+        ssize_t ret = stream->write(data, len);
+        if (ret == -EAGAIN) {
+            continue;
+        } else if (ret < 0) {
+            break;
+        }
+        written += ret;
+        data += ret;
+        len -= ret;
+    }
+
+    printf("> %ld\n", (long) written);
+
+    return written;
+}
+
+static void ppp_link_status(ppp_pcb *pcb, int err_code, void *ctx)
```

Comment:
It's not debug only, it sets state too. Its a callback function when pppos() state changes.